src/persist-client/src/impl/encoding.rs

```diff
@@ -0,0 +1,416 @@
+use timely::PartialOrder;
+use uuid::Uuid;
+
+use crate::error::InvalidUsage;
+use crate::r#impl::state::{
+    HollowBatch, ProtoHollowBatch, ProtoHollowBatchPart, ProtoReader, ProtoSnapshotSplit,
+    ProtoStateRollup, ProtoTrace, ProtoU64Antichain, ProtoU64Description, ReadCapability, State,
+    StateCollections,
+};
+use crate::r#impl::trace::Trace;
+use crate::read::{ReaderId, SnapshotSplit};
+use crate::ShardId;
+
+pub(crate) fn parse_id(id_prefix: char, id_type: &str, x: &str) -> Result<[u8; 16], String> {
+    let u = match x.strip_prefix(id_prefix) {
+        Some(x) => x,
+        None => return Err(format!("invalid {} {}: incorrect prefix", id_type, x)),
+    };
+    let uuid = Uuid::parse_str(&u).map_err(|err| format!("invalid {} {}: {}", id_type, x, err))?;
+    Ok(*uuid.as_bytes())
+}
+
+impl RustType<String> for ShardId {
+    fn into_proto(&self) -> String {
+        self.to_string()
+    }
+
+    fn from_proto(proto: String) -> Result<Self, TryFromProtoError> {
+        match parse_id('s', "ShardId", &proto) {
+            Ok(x) => Ok(ShardId(x)),
+            Err(_) => Err(TryFromProtoError::InvalidShardId(proto)),
+        }
+    }
+}
+
+impl RustType<String> for ReaderId {
+    fn into_proto(&self) -> String {
+        self.to_string()
+    }
+
+    fn from_proto(proto: String) -> Result<Self, TryFromProtoError> {
+        match parse_id('r', "ReaderId", &proto) {
+            Ok(x) => Ok(ReaderId(x)),
+            Err(_) => Err(TryFromProtoError::InvalidShardId(proto)),
+        }
+    }
+}
+
+impl<K, V, T, D> State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    pub fn decode(buf: &[u8]) -> Result<Self, InvalidUsage<T>> {
+        let proto = ProtoStateRollup::decode_checksummed(buf)
+            // We received a State that we couldn't decode. This could happen if
+            // persist messes up backward/forward compatibility, if the durable
+            // data was corrupted, or if operations messes up deployment. In any
+            // case, fail loudly.
+            .expect("internal error: invalid encoded state");
+        Self::try_from(proto).expect("internal error: invalid encoded state")
+    }
+}
+
+impl<K, V, T, D> Codec for State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    fn codec_name() -> String {
+        "md5+protobuf[State]".into()
+    }
+
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: bytes::BufMut,
+    {
+        // TODO: Move checksum to be a field on the proto instead. We can encode
+        // the proto, checksum'ing as we go, and then manually append it onto
+        // the end.
+        //
+        // TODO: Regardless of the above TODO, compute the checksum as we go and
+        // avoid this temp Vec.
+        let temp = self.into_proto().encode_to_vec();
+        buf.put_slice(&temp);
+        let checksum = ProtoStateRollup::md5_checksum(&temp);
```

Comment:
just wanna add this here from https://developers.google.com/protocol-buffers/docs/encoding

---

```
Do not assume the byte output of a serialized message is stable. This is especially true for messages with transitive bytes fields representing other serialized protocol buffer messages.
By default, repeated invocations of serialization methods on the same protocol buffer message instance may not produce the same byte output. That is, the default serialization is not deterministic.

    Deterministic serialization only guarantees the same byte output for a particular binary. The byte output may change across different versions of the binary.
```

---

I _think_ we're still OK doing these checksums, since we just want to verify the integrity of the data written, not compare against other protobufs or repeat encodings of the same data, but wanted to highlight it to be sure.